Reduce the time a Mercedes-Benz spends on the test
bench


Problem Statement Scenario:
Mercedes-Benz engineers have developed a robust testing system, to ensure the safety and reliability of every unique carconfiguration before they hit the road. However, optimizing the speed of their testing system for many possible featurecombinations is complex and time-consuming without a powerful algorithmic approach. Optimal algorithms will contributeto faster testing, resulting in lower carbon dioxide emissions without reducing Daimler’s standards.
Following actions should be performed:
1. If for any column(s), the variance is equal to zero, then you need to remove those variable(s).
2. Check for null and unique values for test and train sets
3. Apply label encoder.
4. Perform dimensionality reduction.
5. Predict your test_df values using xgboost

In [1]:
#Import Libraries requerd for solution
import pandas as pd
import numpy as np
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import LabelEncoder
from math import sqrt

In [2]:
#read the test and train data
train_df = pd.read_csv("mb_train.csv",index_col=0)
test_df = pd.read_csv("mb_test.csv", index_col=0)

In [3]:
train_df.head()

y  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  \
ID                                         ...                                 
0   130.81   k  v  at  a  d  u  j  o    0  ...     0     0     1     0     0   
6    88.53   k  t  av  e  d  y  l  o    0  ...     1     0     0     0     0   
7    76.26  az  w   n  c  d  x  j  x    0  ...     0     0     0     0     0   
9    80.62  az  t   n  f  d  x  l  e    0  ...     0     0     0     0     0   
13   78.02  az  v   n  f  d  h  d  n    0  ...     0     0     0     0     0   

    X380  X382  X383  X384  X385  
ID                                
0      0     0     0     0     0  
6      0     0     0     0     0  
7      0     1     0     0     0  
9      0     0     0     0     0  
13     0     0     0     0     0  

[5 rows x 377 columns]

In [4]:
test_df.head()

X0 X1  X2 X3 X4 X5 X6 X8  X10  X11  ...  X375  X376  X377  X378  X379  \
ID                                      ...                                 
1   az  v   n  f  d  t  a  w    0    0  ...     0     0     0     1     0   
2    t  b  ai  a  d  b  g  y    0    0  ...     0     0     1     0     0   
3   az  v  as  f  d  a  j  j    0    0  ...     0     0     0     1     0   
4   az  l   n  f  d  z  l  n    0    0  ...     0     0     0     1     0   
5    w  s  as  c  d  y  i  m    0    0  ...     1     0     0     0     0   

    X380  X382  X383  X384  X385  
ID                                
1      0     0     0     0     0  
2      0     0     0     0     0  
3      0     0     0     0     0  
4      0     0     0     0     0  
5      0     0     0     0     0  

[5 rows x 376 columns]

Data Evaluation

In [5]:
train_df.head()

y  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  \
ID                                         ...                                 
0   130.81   k  v  at  a  d  u  j  o    0  ...     0     0     1     0     0   
6    88.53   k  t  av  e  d  y  l  o    0  ...     1     0     0     0     0   
7    76.26  az  w   n  c  d  x  j  x    0  ...     0     0     0     0     0   
9    80.62  az  t   n  f  d  x  l  e    0  ...     0     0     0     0     0   
13   78.02  az  v   n  f  d  h  d  n    0  ...     0     0     0     0     0   

    X380  X382  X383  X384  X385  
ID                                
0      0     0     0     0     0  
6      0     0     0     0     0  
7      0     1     0     0     0  
9      0     0     0     0     0  
13     0     0     0     0     0  

[5 rows x 377 columns]

In [6]:
test_df.head()

X0 X1  X2 X3 X4 X5 X6 X8  X10  X11  ...  X375  X376  X377  X378  X379  \
ID                                      ...                                 
1   az  v   n  f  d  t  a  w    0    0  ...     0     0     0     1     0   
2    t  b  ai  a  d  b  g  y    0    0  ...     0     0     1     0     0   
3   az  v  as  f  d  a  j  j    0    0  ...     0     0     0     1     0   
4   az  l   n  f  d  z  l  n    0    0  ...     0     0     0     1     0   
5    w  s  as  c  d  y  i  m    0    0  ...     1     0     0     0     0   

    X380  X382  X383  X384  X385  
ID                                
1      0     0     0     0     0  
2      0     0     0     0     0  
3      0     0     0     0     0  
4      0     0     0     0     0  
5      0     0     0     0     0  

[5 rows x 376 columns]

## Task1: If for any column(s), the variance is equal to zero, then you need to remove those variable(s).

In [7]:
zeroVarCol = train_df.var()[train_df.var()==0].index.values
zeroVarCol

C:\Users\shubh\AppData\Local\Temp/ipykernel_33260/2735335373.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  zeroVarCol = train_df.var()[train_df.var()==0].index.values


array(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], dtype=object)

In [8]:
train_df = train_df.drop(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], axis = 1)
test_df = test_df.drop(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], axis = 1)

In [9]:
test_df.shape

(4209, 364)

In [10]:
train_df.shape

(4209, 365)

## Task 2: Check for null and unique values for test and train sets

In [11]:
#checking for null values
np.sum(train_df.isnull().sum())

0

In [12]:
np.sum(test_df.isnull().sum())

0

In [13]:
#checking for unique values
np.sum(train_df.nunique().sum())

3452

In [14]:
np.sum(test_df.nunique().sum())

908

In [15]:
#checking for duplicates
np.sum(train_df.duplicated().sum())

1

In [16]:
np.sum(test_df.duplicated().sum())

299

## Task 3: Apply label encoder.

In [17]:
objectColumns = train_df.describe(include=[object]).columns.values
objectColumns

array(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype=object)

In [18]:
labelEcoder = LabelEncoder()
for x in objectColumns:
    labelEcoder.fit(train_df[x].append(test_df[x]).values)
    train_df[x] = labelEcoder.transform(train_df[x])
    test_df[x] = labelEcoder.transform(test_df[x])

In [19]:
X = train_df.drop(["y"], axis=1)
y = train_df["y"]

from sklearn import model_selection
X_train, X_test, y_train, y_test = model_selection.train_test_split(X,y,test_size=0.3, random_state=1)

## Task 4: Perform dimensionality reduction. 

In [20]:
from sklearn.decomposition import PCA as sPCA
pca = sPCA(.98, svd_solver="full")
sklearn_pca = pca.fit(X)

In [21]:
sklearn_pca.n_components_

12

Using Principal component analysis (PCA), decomposition of the data i.e., 364 components with 98% variance toproject it to a lower dimensional space is 12 components

In [22]:
sklearn_pca.explained_variance_ratio_

array([0.40868988, 0.21758508, 0.13120081, 0.10783522, 0.08165248,
       0.0140934 , 0.00660951, 0.00384659, 0.00260289, 0.00214378,
       0.00209857, 0.00180388])

## Task 5. Predict your test_df values using xgboost


In [23]:
pca_xtrain = pd.DataFrame(sklearn_pca.transform(X_train))
pca_xtest = pd.DataFrame(sklearn_pca.transform(X_test))
pca_df_test_data = pd.DataFrame(sklearn_pca.transform(test_df))

In [24]:
import xgboost as xgb
xgb_regressor_model = xgb.XGBRegressor(objective = 'reg:squarederror', learning_rate= 0.1 )
xgb_regressor_model.fit(pca_xtrain, y_train)

XGBRegressor(base_score=0.5, booster='gbtree', callbacks=None,
             colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
             early_stopping_rounds=None, enable_categorical=False,
             eval_metric=None, gamma=0, gpu_id=-1, grow_policy='depthwise',
             importance_type=None, interaction_constraints='',
             learning_rate=0.1, max_bin=256, max_cat_to_onehot=4,
             max_delta_step=0, max_depth=6, max_leaves=0, min_child_weight=1,
             missing=nan, monotone_constraints='()', n_estimators=100, n_jobs=0,
             num_parallel_tree=1, predictor='auto', random_state=0, reg_alpha=0,
             reg_lambda=1, ...)

In [25]:
xgb_regressor_model_predicted_y_test = xgb_regressor_model.predict(pca_xtest)
print(sqrt(mean_squared_error(y_test, xgb_regressor_model_predicted_y_test)))

8.555776313214766


In [26]:
xgb_RFregressor_model = xgb.XGBRFRegressor(objective = 'reg:squarederror', learning_rate= 1)
xgb_RFregressor_model.fit(pca_xtrain, y_train)

XGBRFRegressor(base_score=0.5, booster='gbtree', callbacks=None,
               colsample_bylevel=1, colsample_bytree=1,
               early_stopping_rounds=None, enable_categorical=False,
               eval_metric=None, gamma=0, gpu_id=-1, grow_policy='depthwise',
               importance_type=None, interaction_constraints='',
               learning_rate=1, max_bin=256, max_cat_to_onehot=4,
               max_delta_step=0, max_depth=6, max_leaves=0, min_child_weight=1,
               missing=nan, monotone_constraints='()', n_estimators=100,
               n_jobs=0, num_parallel_tree=100, objective='reg:squarederror',
               predictor='auto', random_state=0, reg_alpha=0,
               sampling_method='uniform', ...)

In [27]:
xgb_RFregressor_model_predicted_y_test = xgb_RFregressor_model.predict(pca_xtest)
print(sqrt(mean_squared_error(y_test, xgb_RFregressor_model_predicted_y_test)))

9.064875902434391


**Comparing both XGBRFregressor provides more acuracy as compare to the XGBregressor**